lib/collection/src/shards/shard_holder.rs

```diff
@@ -155,6 +206,52 @@ impl ShardHolder {
         Ok(())
     }
 
+    pub fn start_resharding(
+        &mut self,
+        shard_id: ShardId,
+        shard: ShardReplicaSet,
+        shard_key: Option<ShardKey>,
+    ) -> Result<(), CollectionError> {
+        let Some(ring) = self.rings.get_mut(&shard_key) else {
+            // TODO: `CollectionError::service_error`? 🤔
+            return Err(CollectionError::bad_request(
+                "shard holder does not contain default hashring".into(),
+            ));
+        };
+
+        if ring.is_resharding() {
+            debug_assert!(
+                self.resharding.is_some(),
+                "shard holder contains resharding hashring, but resharding field is {:?}",
+                self.resharding
+            );
+
+            // TODO: `CollectionError::service_error`? 🤔
+            return Err(CollectionError::bad_request(
+                "shard holder already contains resharding hashring".into(),
+            ));
+        }
+
+        debug_assert!(
+            self.resharding.is_none(),
+            "shard holder does not contain resharding hashring, but resharding field is {:?}",
+            self.resharding
+        );
+
+        if self.shards.contains_key(&shard_id) {
+            // TODO: `CollectionError::service_error`? 🤔
+            return Err(CollectionError::bad_request(format!(
+                "shard holder already contains shard {shard_id} replica set"
+            )));
```

Comment:
Yes, a service error is better in all of these cases. It's something internal.

Reply:
I kinda agree, but the trick is that `start_resharding` is called by consensus thread (when `Resharding::Start` message is applied). And if `start_resharding` returns *service* error, it will make consensus thread to panic/exit.

So I'm not *yet* 100% sure if any/some/all of these errors can be somehow safeguarded on another level (e.g., some validation when initially proposing `Resharding::Start`, or something).

The first two, most likely, can be made service errors, but I'm not sure about the third one. I propose to leave it as-is for now, and once we implement more of resharding, 1) we might restructure the code so these error conditions aren't there at all anymore, or 2) we get better understanding, which conditions can or cannot *ever* happen.

Reply:
> it will make consensus thread to panic/exit.

:disappointed: